### PR TITLE
fix: HoldingAmfUe restore when CM-Idle

### PR DESCRIPTION
**Pull request by andy89923**
## Description 
This PR fixes the issue that if the UE is in CM-Idle state, the UE initiating ServiceRequest will require authentication again. 

## Others
- Related PR #136 
  - Rename `FindAmfUE` to `HoldingAmfUe`

### CODE REVIEW

**Comment by andy89923** [comment]
Hi @ss920386 

Could you please help to review this PR?

Tks.

File: internal/nas/handler.go [hunk 1]
```diff
@@ -10,39 +10,49 @@ import (
...
 		return
 	}
 
-	if ue.AmfUe == nil {
-		ue.AmfUe = amfSelf.NewAmfUe("")
-		gmm_common.AttachRanUeToAmfUeAndReleaseOldIfAny(ue.AmfUe, ue)
+	if ranUe.AmfUe == nil {
+		if ranUe.FindAmfUe != nil && !ranUe.FindAmfUe.CmConnect(ranUe.Ran.AnType) {
+			// models.CmState_IDLE
+			gmm_common.ClearHoldingRanUe(ranUe.FindAmfUe.RanUe[ranUe.Ran.AnType])
```
Comment (ianchen0119) [inline internal/nas/handler.go:29]: redundant.

File: internal/nas/handler.go [hunk 2]
```diff
@@ -10,39 +10,49 @@ import (
...
+
+			ranUe.AmfUe = ranUe.FindAmfUe
```
Comment (ianchen0119) [inline internal/nas/handler.go:31]: redundant.

File: internal/nas/handler.go [hunk 3]
```diff
@@ -10,39 +10,49 @@ import (
...
+			gmm_common.AttachRanUeToAmfUeAndReleaseOldIfAny(ranUe.AmfUe, ranUe)
+			ranUe.FindAmfUe = nil
+		} else {
+			// New AmfUe
+			ranUe.AmfUe = amfSelf.NewAmfUe("")
```
Comment (ianchen0119) [inline internal/nas/handler.go:36]: The case of Service Request
UE context is in CM-IDLE state.

File: internal/nas/handler.go [hunk 4]
```diff
@@ -10,39 +10,49 @@ import (
...
+			gmm_common.AttachRanUeToAmfUeAndReleaseOldIfAny(ranUe.AmfUe, ranUe)
+		}
 	}
 
-	msg, integrityProtected, err := nas_security.Decode(ue.AmfUe, ue.Ran.AnType, nasPdu, initialMessage)
+	msg, integrityProtected, err := nas_security.Decode(ranUe.AmfUe, ranUe.Ran.AnType, nasPdu, initialMessage)
 	if err != nil {
-		ue.AmfUe.NASLog.Errorln(err)
+		ranUe.AmfUe.NASLog.Errorln(err)
```
Comment (ianchen0119) [inline internal/nas/handler.go:43]: Assume that we have an existing UE context, which is in CM-CONNECTED state.
We will release it if the new UE context has a valid security context in line 53.

File: internal/nas/handler.go (hunk 1 above)
Comment (ss920386) [inline internal/nas/handler.go:29]: ` !ranUe.FindAmfUe.CmConnect(ranUe.Ran.AnType)` implies that ranUe.FinAmfUe didn't have underlying ranUe, right? 
Shouldn't ClearHoldingRanUe(ranUe.FindAmfUe.RanUe[ranUe.Ran.AnType]) do nothing?

Comment (andy89923) [reply internal/nas/handler.go]: The `ClearHoldingRanUe(ranUe.FindAmfUe.RanUe[ranUe.Ran.AnType])` do noting. 
This line had been remove in the latest commit.

**Review by ianchen0119 — APPROVED** [review]

**Review by ss920386 — APPROVED** [review]

**Comment by andy89923** [comment]
Thanks a lot. @ianchen0119 & @ss920386

**Pull request merged**